### ISSUE TITLE
[Documentation] Supported ruby-libraries, in particular compiled ones such as ruby-gtk?

### ISSUE THREAD

**Issue by rubyFeedback** [jruby/jruby#6561]
Hey jruby folks,

This is a documentation request.

I will explain it and the background.

I have been using ruby since ~2005 or so. I have written lots of ruby code.

I have used jruby in the past but I did not use it "seriously", mostly because
I am still a java noob. I write a bit java here and there but I am mostly a ruby
guy.

Anyway. I have been experimenting a LOT with ruby-gtk3 and I'd love to
try my projects in java / jruby too, as I learn or get better with java as well.

Coming from this point of view, I'd like to see some document, perhaps
a. md file, or wiki entry, whichever way fits, to list which of the compiled
libraries would work on jruby.

Now, ruby-gtk3 works here as-is via ruby, so we can say jruby would
work 1:1 just fine too BUT my question is also aimed at whether java
would make the net-result faster. I don't care about start-up time that
much, but once running, I'd like to offer this all as fast as possible,
and I think java is easier to work with than C and C++, so I rather want
to give that a more serious try first. (Prototyping I would still do with
ruby as such, but perhaps via jruby I could have a widget toolset
in gtk that would be faster than using pure ruby only)

**Comment by headius** [comment]
Sadly, it appears that all the Ruby GTK libraries use CRuby native extensions, which are not supported on JRuby. The APIs required are invasive and impossible to make thread-safe, so JRuby users generally use JVM libraries (callable from Ruby like any Ruby class) or FFI if calls to a native lib are needed.

This probably isn't what you want to hear, but there are many GUI libraries for JRuby that could be used instead. Nearly all of them require no native library or additional builds on the target system. A short list of options:

* Swing, the old classic Java GUI API that works everywhere
* JavaFX, a more modern scene-based API with a large library of widgets and great performance
* SWT, a wrapper around native GUI APIs (including, I believe, GTK) but with a single standard API and good cross-platform support.

JRuby apps can also be bundled up as a single file with all dependencies and shipped to any system with a JVM... no other packages are required.

Bottom line is that any app using the Ruby GTK bindings will be unable to run on JRuby.

**Comment by headius** [comment]
FWIW, you don't have to know anything about Java to start using JRuby in general, but a little Java knowledge helps once you start calling out to JVM libraries like the ones I listed.

We would be happy to chat with you about using JRuby for GUI projects! Feel free to stop by our Matrix chat:

https://www.jruby.org/chat
